x-pack/plugins/elastic_assistant/server/routes/helpers.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { KibanaRequest } from '@kbn/core-http-server';
+import { Logger } from '@kbn/core/server';
+import { RequestBody } from '../lib/langchain/types';
+
+interface GetPluginNameFromRequestParams {
+  request: KibanaRequest<unknown, unknown, RequestBody>;
+  logger?: Logger;
+}
+
+export const DEFAULT_PLUGIN_NAME = 'securitySolutionUI';
+
+/**
+ * Attempts to extract the plugin name the request originated from using the request headers.
+ *
+ * Note from Kibana Core: This is not a 100% fit solution, though, because plugins can run in the background,
+ * or even use other plugins’ helpers (ie, APM can use the infra helpers to call a third plugin)
+ *
+ * Should suffice for our purposes here with where the Elastic Assistant is currently used, but if needing a
+ * dedicated solution, the core folks said to reach out.
+ *
+ * @param logger optional logger to log any errors
+ * @param request Kibana Request
+ *
+ * @returns plugin name
+ */
+export const getPluginNameFromRequest = ({
+  logger,
+  request,
+}: GetPluginNameFromRequestParams): string => {
+  try {
+    const contextHeader = request.headers['x-kbn-context'];
+    if (contextHeader != null) {
+      return JSON.parse(
+        decodeURIComponent(Array.isArray(contextHeader) ? contextHeader[0] : contextHeader)
+      )?.name;
+    }
+  } catch (err) {
+    logger?.error('Error determining source plugin, using default assistant tools.');
+  }
+  return DEFAULT_PLUGIN_NAME;
```

Comment:
Consider making `defaultPluginName` a required parameter to `getPluginNameFromRequest()`, because this would require the caller to specify the fallback case, making it unnecessary to always return `securitySolutionUI` as the fallback.

Reply:
Addressed in [70186ab](https://github.com/elastic/kibana/pull/172234/commits/70186ab4c837c88bb6e3d99210ebf91c9a402369)